common/nym-kcp/src/session.rs

```diff
@@ -0,0 +1,1778 @@
+
+        // In real KCP, we check against the number of unacknowledged segments:
+        // while self.snd_nxt < self.snd_una + cwnd { ... }
+        // Here, we approximate by checking the current length of snd_buf against cwnd.
+        while let Some(mut seg) = self.snd_queue.pop_front() {
+            // Check if adding this segment would exceed the congestion window
+            if (self.snd_buf.len() as u16) >= cwnd {
+                // Effective window is full
+                self.snd_queue.push_front(seg); // Put it back
+                break;
+            }
+            // init rto
+            seg.xmit = 0;
+            seg.rto = self.rx_rto;
+            seg.resendts = 0; // will set later
+            seg.ts = self.current;
+            self.snd_buf.push_back(seg);
+        }
+    }
+
+    /// Build KcpPacket(s) for segments needing send or retransmit.
+    fn flush_outgoing(&mut self) {
+        // Log current snd_buf state before iterating
+        // let current_sns: Vec<u32> = self.snd_buf.iter().map(|seg| seg.sn).collect();
+        // debug!(
+        //     "[ConvID: {}, Thread: {:?}] flush_outgoing: Checking snd_buf (len={}): {:?}",
+        //     self.conv,
+        //     thread::current().id(),
+        //     self.snd_buf.len(),
+        //     current_sns
+        // );
+
+        for seg in &mut self.snd_buf {
+            let mut need_send = false;
+            if seg.xmit == 0 {
+                // never sent
+                need_send = true;
+                seg.xmit = 1;
+                seg.resendts = self.current + seg.rto;
+            } else if self.current >= seg.resendts {
+                // time to retransmit
+                need_send = true;
+                seg.xmit += 1;
+                // Exponential backoff: double RTO for this segment
+                seg.rto *= 2;
+                // Clamp to the session's maximum RTO (hardcoded as 60s for now)
+                const MAX_RTO: u32 = 60000; // Same as used in update_rtt
+                if seg.rto > MAX_RTO {
+                    seg.rto = MAX_RTO;
+                }
+                seg.resendts = self.current + seg.rto;
+                info!(
+                    "{}",
+                    Yellow.paint(format!(
+                        "Retransmit conv_id: {}, sn: {}, frg: {}",
+                        self.conv, seg.sn, seg.frg
+                    ))
+                );
+            }
+
+            if need_send {
+                // Make a KcpPacket
+                let pkt = KcpPacket::new(
+                    self.conv,
+                    KcpCommand::Push,
+                    seg.frg,
+                    self.rcv_wnd,
+                    seg.ts, // original send timestamp
+                    seg.sn,
+                    self.rcv_nxt, // self.rcv_nxt for ack
+                    seg.data.clone(),
+                );
+                self.out_pkts.push(pkt);
+
+                // if too many xmit => dead_link check, etc.
```

Comment:
so that's still to be done, right?

Reply:
That was supposed to come when we implement backpressure, we just never got to it